<!doctype html>
<html class="no-js" lang="en" data-content_root="../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="prev" title="singler" href="modules.html" />

    <!-- Generated with Sphinx 8.2.3 and Furo 2024.08.06 -->
        <title>singler package - singler unknown documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=354aac6f" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=302659d7" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">singler unknown documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  
  <span class="sidebar-brand-text">singler unknown documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../readme.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributions &amp; Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="../license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Changelog</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="modules.html">Module Reference</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Module Reference</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">singler package</a></li>
</ul>
</li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../_sources/api/singler.rst.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="module-singler">
<span id="singler-package"></span><h1>singler package<a class="headerlink" href="#module-singler" title="Link to this heading">¶</a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading">¶</a></h2>
</section>
<section id="module-singler.aggregate_reference">
<span id="singler-aggregate-reference-module"></span><h2>singler.aggregate_reference module<a class="headerlink" href="#module-singler.aggregate_reference" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="singler.aggregate_reference.aggregate_reference">
<span class="sig-prename descclassname"><span class="pre">singler.aggregate_reference.</span></span><span class="sig-name descname"><span class="pre">aggregate_reference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_centers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">power</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_top</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rank</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'logcounts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset_row</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/aggregate_reference.html#aggregate_reference"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.aggregate_reference.aggregate_reference" title="Link to this definition">¶</a></dt>
<dd><p>Aggregate reference samples for a given label by using vector quantization to average their count profiles.
The idea is to reduce the size of single-cell reference datasets so as to reduce the computation time of <a class="reference internal" href="#singler.train_single.train_single" title="singler.train_single.train_single"><code class="xref py py-func docutils literal notranslate"><span class="pre">train_single()</span></code></a>.
We perform k-means clustering for all cells in each label and aggregate all cells within each k-means cluster.
(More specifically, the clustering is done on the principal components generated from the highly variable genes to better capture the structure within each label.)
This yields one or more profiles per label, reducing the number of separate observations while preserving some level of intra-label heterogeneity.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref_data</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Any" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – Floating-point matrix of reference expression values, usually containing log-expression values.
Alternatively, a <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.5)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> object containing such a matrix.</p></li>
<li><p><strong>ref_labels</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Array of length equal to the number of columns in <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>, containing the labels for each cell.</p></li>
<li><p><strong>ref_features</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Sequence of identifiers for each feature, i.e., row in <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>.</p></li>
<li><p><strong>num_centers</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Optional" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Maximum number of aggregated profiles to produce for each label with <code class="xref py py-func docutils literal notranslate"><span class="pre">cluster_kmeans()</span></code>.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, a suitable number of profiles is automatically chosen.</p></li>
<li><p><strong>power</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#float" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Number between 0 and 1 indicating how much aggregation should be performed.
Specifically, we set the number of clusters to <code class="docutils literal notranslate"><span class="pre">X**power</span></code> where <code class="docutils literal notranslate"><span class="pre">X</span></code> is the number of cells assigned to that label.
Ignored if <code class="docutils literal notranslate"><span class="pre">num_centers</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>num_top</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of highly variable genes to use for PCA prior to clustering, see <code class="xref py py-func docutils literal notranslate"><span class="pre">choose_highly_variable_genes()</span></code>.</p></li>
<li><p><strong>rank</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of principal components to use during clustering, see <code class="xref py py-func docutils literal notranslate"><span class="pre">run_pca()</span></code>.</p></li>
<li><p><strong>assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</span>) – Integer or string specifying the assay of <code class="docutils literal notranslate"><span class="pre">ref_data</span></code> containing the relevant expression matrix,
if <code class="docutils literal notranslate"><span class="pre">ref</span></code> is a <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.5)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> object.</p></li>
<li><p><strong>subset_row</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Optional" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Array of row indices specifying the rows of <code class="docutils literal notranslate"><span class="pre">ref_data</span></code> to use for clustering.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, no additional filtering is performed.
Note that even if <code class="docutils literal notranslate"><span class="pre">subset_row</span></code> is provided, aggregation is still performed on all genes.</p></li>
<li><p><strong>check_missing</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to check for and remove rows with missing (NaN) values from <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.5)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.5)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> containing the aggregated values in its first assay.
The label for each aggregated profile is stored in the column data.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-singler.annotate_integrated">
<span id="singler-annotate-integrated-module"></span><h2>singler.annotate_integrated module<a class="headerlink" href="#module-singler.annotate_integrated" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="singler.annotate_integrated.annotate_integrated">
<span class="sig-prename descclassname"><span class="pre">singler.annotate_integrated.</span></span><span class="sig-name descname"><span class="pre">annotate_integrated</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">test_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'logcounts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_check_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_check_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_single_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">classify_single_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_integrated_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">classify_integrated_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/annotate_integrated.html#annotate_integrated"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.annotate_integrated.annotate_integrated" title="Link to this definition">¶</a></dt>
<dd><p>Annotate a single-cell expression dataset based on the correlation
of each cell to profiles in multiple labelled references, where the
annotation from each reference is then integrated across references.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>test_data</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Any" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – <p>A matrix-like object representing the test dataset, where rows are
features and columns are samples (usually cells). Entries should be
expression values; only the ranking within each column is used.</p>
<p>Alternatively, a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.5)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>
containing such a matrix in one of its assays.</p>
</p></li>
<li><p><strong>ref_data</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – <p>Sequence consisting of one or more of the following:</p>
<ul>
<li><p>A matrix-like object representing the reference dataset, where rows
are features and columns are samples. Entries should be expression values,
usually log-transformed (see comments for the <code class="docutils literal notranslate"><span class="pre">ref_data</span></code> argument in
<a class="reference internal" href="#singler.train_single.train_single" title="singler.train_single.train_single"><code class="xref py py-func docutils literal notranslate"><span class="pre">train_single()</span></code></a>).</p></li>
<li><p>A <code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code> object containing such a matrix in its assays.</p></li>
</ul>
</p></li>
<li><p><strong>ref_labels</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#list" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Sequence of the same length as <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>. The <code class="docutils literal notranslate"><span class="pre">i</span></code>-th entry
should be a sequence of length equal to the number of columns of
<code class="docutils literal notranslate"><span class="pre">ref_data[i]</span></code>, containing the label associated with each column.</p></li>
<li><p><strong>test_features</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Optional" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Sequence of length equal to the number of rows in <code class="docutils literal notranslate"><span class="pre">test_data</span></code>,
containing the feature identifier for each row.  Alternatively
<code class="docutils literal notranslate"><span class="pre">None</span></code>, to use the row names of the experiment as features.</p></li>
<li><p><strong>ref_features</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Optional" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#list" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Optional" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]]]</span>) – <p>Sequence of the same length as <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>. The <code class="docutils literal notranslate"><span class="pre">i</span></code>-th entry
should be a sequence of length equal to the number of rows of
<code class="docutils literal notranslate"><span class="pre">ref_data[i]</span></code>, containing the feature identifier associated with
each row. It can also be set to <code class="docutils literal notranslate"><span class="pre">None</span></code> to use the row names of
the experiment as features.</p>
<p>This can also be <code class="docutils literal notranslate"><span class="pre">None</span></code> to indicate that the row names should be
used for all references, assuming <code class="docutils literal notranslate"><span class="pre">ref_data</span></code> only contains
<code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code> objects.</p>
</p></li>
<li><p><strong>test_assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Assay of <code class="docutils literal notranslate"><span class="pre">test_data</span></code> containing the expression matrix, if <code class="docutils literal notranslate"><span class="pre">test_data</span></code> is a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.5)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>.</p></li>
<li><p><strong>test_check_missing</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to check for and remove missing (i.e., NaN) values from the test dataset.</p></li>
<li><p><strong>ref_assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Assay containing the expression matrix for any entry of <code class="docutils literal notranslate"><span class="pre">ref_data</span></code> that is a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.5)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>.</p></li>
<li><p><strong>ref_check_missing</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to check for and remove missing (i.e., NaN) values from the reference datasets.</p></li>
<li><p><strong>train_single_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Further arguments to pass to
<a class="reference internal" href="#singler.train_single.train_single" title="singler.train_single.train_single"><code class="xref py py-func docutils literal notranslate"><span class="pre">train_single()</span></code></a>.</p></li>
<li><p><strong>classify_single_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Further arguments to pass to
<a class="reference internal" href="#singler.classify_single.classify_single" title="singler.classify_single.classify_single"><code class="xref py py-func docutils literal notranslate"><span class="pre">classify_single()</span></code></a>.</p></li>
<li><p><strong>train_integrated_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Further arguments to pass to
<a class="reference internal" href="#singler.train_integrated.train_integrated" title="singler.train_integrated.train_integrated"><code class="xref py py-func docutils literal notranslate"><span class="pre">train_integrated()</span></code></a>.</p></li>
<li><p><strong>classify_integrated_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Further arguments to pass to
<a class="reference internal" href="#singler.classify_integrated.classify_integrated" title="singler.classify_integrated.classify_integrated"><code class="xref py py-func docutils literal notranslate"><span class="pre">classify_integrated()</span></code></a>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use for the various steps.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Tuple" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#list" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>[<a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>], <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>]</span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple where the first element contains per-reference results (i.e. a
list of BiocFrame outputs, roughly equivalent to running
<a class="reference internal" href="#singler.annotate_single.annotate_single" title="singler.annotate_single.annotate_single"><code class="xref py py-func docutils literal notranslate"><span class="pre">annotate_single()</span></code></a> on each reference)
and the second element contains integrated results across references
(i.e., a BiocFrame from
<a class="reference internal" href="#singler.classify_integrated.classify_integrated" title="singler.classify_integrated.classify_integrated"><code class="xref py py-func docutils literal notranslate"><span class="pre">classify_integrated()</span></code></a>).</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-singler.annotate_single">
<span id="singler-annotate-single-module"></span><h2>singler.annotate_single module<a class="headerlink" href="#module-singler.annotate_single" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="singler.annotate_single.annotate_single">
<span class="sig-prename descclassname"><span class="pre">singler.annotate_single.</span></span><span class="sig-name descname"><span class="pre">annotate_single</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">test_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_check_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_check_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">classify_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/annotate_single.html#annotate_single"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.annotate_single.annotate_single" title="Link to this definition">¶</a></dt>
<dd><p>Annotate a single-cell expression dataset based on the correlation
of each cell to profiles in a labelled reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>test_data</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Any" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – <p>A matrix-like object representing the test dataset, where rows are
features and columns are samples (usually cells). Entries should be expression
values; only the ranking within each column will be used.</p>
<p>Alternatively, a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.5)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>
containing such a matrix in one of its assays. Non-default assay
types can be specified in <code class="docutils literal notranslate"><span class="pre">classify_args</span></code>.</p>
</p></li>
<li><p><strong>ref_data</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Any" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – <p>A matrix-like object representing the reference dataset, where rows
are features and columns are samples. Entries should be expression values,
usually log-transformed (see comments for the <code class="docutils literal notranslate"><span class="pre">ref</span></code> argument in
<a class="reference internal" href="#singler.train_single.train_single" title="singler.train_single.train_single"><code class="xref py py-func docutils literal notranslate"><span class="pre">train_single()</span></code></a>).</p>
<p>Alternatively, a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.5)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>
containing such a matrix in one of its assays. Non-default assay
types can be specified in <code class="docutils literal notranslate"><span class="pre">classify_args</span></code>.</p>
</p></li>
<li><p><strong>ref_labels</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Sequence of length equal to the number of columns of <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>,
containing the label associated with each column.</p></li>
<li><p><strong>test_features</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Optional" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Sequence of length equal to the number of rows in <code class="docutils literal notranslate"><span class="pre">test_data</span></code>,
containing the feature identifier for each row. Alternatively
<code class="docutils literal notranslate"><span class="pre">None</span></code>, to use the row names of the experiment as features.</p></li>
<li><p><strong>ref_features</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Optional" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Sequence of length equal to the number of rows of <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>,
containing the feature identifier for each row. Alternatively
<code class="docutils literal notranslate"><span class="pre">None</span></code>, to use the row names of the experiment as features.</p></li>
<li><p><strong>test_assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Assay containing the expression matrix, if <code class="docutils literal notranslate"><span class="pre">test_data</span></code> is a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.5)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>.</p></li>
<li><p><strong>ref_assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Assay containing the expression matrix, if <code class="docutils literal notranslate"><span class="pre">ref_data</span></code> is a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.5)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>.</p></li>
<li><p><strong>test_assay_type</strong> – Whether to remove rows with missing values from the test dataset.</p></li>
<li><p><strong>ref_assay_type</strong> – Whether to remove rows with missing values from the reference dataset.</p></li>
<li><p><strong>train_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Further arguments to pass to
<a class="reference internal" href="#singler.train_single.train_single" title="singler.train_single.train_single"><code class="xref py py-func docutils literal notranslate"><span class="pre">train_single()</span></code></a>.</p></li>
<li><p><strong>classify_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Further arguments to pass to
<a class="reference internal" href="#singler.classify_single.classify_single" title="singler.classify_single.classify_single"><code class="xref py py-func docutils literal notranslate"><span class="pre">classify_single()</span></code></a>.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use for the various steps.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> of labelling results, see
<a class="reference internal" href="#singler.classify_single.classify_single" title="singler.classify_single.classify_single"><code class="xref py py-func docutils literal notranslate"><span class="pre">classify_single()</span></code></a> for details.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-singler.classify_integrated">
<span id="singler-classify-integrated-module"></span><h2>singler.classify_integrated module<a class="headerlink" href="#module-singler.classify_integrated" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="singler.classify_integrated.classify_integrated">
<span class="sig-prename descclassname"><span class="pre">singler.classify_integrated.</span></span><span class="sig-name descname"><span class="pre">classify_integrated</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">test_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">results</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">integrated_prebuilt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quantile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_fine_tune</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fine_tune_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/classify_integrated.html#classify_integrated"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.classify_integrated.classify_integrated" title="Link to this definition">¶</a></dt>
<dd><p>Integrate classification results across multiple references for a single test dataset.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>test_data</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Any" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – <p>A matrix-like object where each row is a feature and each column
is a test sample (usually a single cell), containing expression values.
Normalized and/or transformed expression values are also acceptable as only
the ranking is used within this function.</p>
<p>Alternatively, a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.5)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>
containing such a matrix in one of its assays.</p>
</p></li>
<li><p><strong>results</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#list" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>[<a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>]</span>) – List of classification results generated by running
<a class="reference internal" href="#singler.classify_single.classify_single" title="singler.classify_single.classify_single"><code class="xref py py-func docutils literal notranslate"><span class="pre">classify_single()</span></code></a> on
<code class="docutils literal notranslate"><span class="pre">test_data</span></code> with each reference. References should be in the
same order as that used to construct <code class="docutils literal notranslate"><span class="pre">integrated_prebuilt</span></code>.</p></li>
<li><p><strong>integrated_prebuilt</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#singler.train_integrated.TrainedIntegratedReferences" title="singler.train_integrated.TrainedIntegratedReferences"><code class="xref py py-class docutils literal notranslate"><span class="pre">TrainedIntegratedReferences</span></code></a></span>) – Integrated reference object, constructed with
<a class="reference internal" href="#singler.train_integrated.train_integrated" title="singler.train_integrated.train_integrated"><code class="xref py py-func docutils literal notranslate"><span class="pre">train_integrated()</span></code></a>.</p></li>
<li><p><strong>assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Assay containing the expression matrix, if <code class="docutils literal notranslate"><span class="pre">test_data</span></code> is a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.5)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>.</p></li>
<li><p><strong>quantile</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#float" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Quantile of the correlation distribution for computing the score for each label.
Larger values increase sensitivity of matches at the expense of
similarity to the average behavior of each label.</p></li>
<li><p><strong>use_fine_tune</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether fine-tuning should be performed. This improves accuracy for
distinguishing between references with similar best labels but
requires more computational work.</p></li>
<li><p><strong>fine_tune_threshold</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#float" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Maximum difference from the maximum correlation to use in
fine-tuning. All references above this threshold are used for
another round of fine-tuning.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use during classification.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a> containing the
<code class="docutils literal notranslate"><span class="pre">best_label</span></code> across all references, defined as the assigned label in
the best reference; the identity of the <code class="docutils literal notranslate"><span class="pre">best_reference</span></code>, either as a
name string or an integer index; the <code class="docutils literal notranslate"><span class="pre">scores</span></code> for the best label in
each reference, as a nested <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code>; and the <code class="docutils literal notranslate"><span class="pre">delta</span></code> from the
best to the second-best reference. Each row corresponds to a column of
<code class="docutils literal notranslate"><span class="pre">test_data</span></code>.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-singler.classify_single">
<span id="singler-classify-single-module"></span><h2>singler.classify_single module<a class="headerlink" href="#module-singler.classify_single" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="singler.classify_single.classify_single">
<span class="sig-prename descclassname"><span class="pre">singler.classify_single.</span></span><span class="sig-name descname"><span class="pre">classify_single</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">test_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_prebuilt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quantile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_fine_tune</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fine_tune_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/classify_single.html#classify_single"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.classify_single.classify_single" title="Link to this definition">¶</a></dt>
<dd><p>Classify a test dataset against a reference by assigning labels from the latter to each column of the former
using the SingleR algorithm.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>test_data</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Any" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – <p>A matrix-like object where each row is a feature and each column
is a test sample (usually a single cell), containing expression values.
Normalized and transformed expression values are also acceptable as only
the ranking is used within this function.</p>
<p>Alternatively, a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.5)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>
containing such a matrix in one of its assays.</p>
</p></li>
<li><p><strong>ref_prebuilt</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#singler.train_single.TrainedSingleReference" title="singler.train_single.TrainedSingleReference"><code class="xref py py-class docutils literal notranslate"><span class="pre">TrainedSingleReference</span></code></a></span>) – A pre-built reference created with
<a class="reference internal" href="#singler.train_single.train_single" title="singler.train_single.train_single"><code class="xref py py-func docutils literal notranslate"><span class="pre">train_single()</span></code></a>.</p></li>
<li><p><strong>assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Assay containing the expression matrix, if <code class="docutils literal notranslate"><span class="pre">test_data</span></code> is a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.5)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>.</p></li>
<li><p><strong>quantile</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#float" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Quantile of the correlation distribution for computing the score for each label.
Larger values increase sensitivity of matches at the expense of
similarity to the average behavior of each label.</p></li>
<li><p><strong>use_fine_tune</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether fine-tuning should be performed. This improves accuracy for distinguishing
between similar labels but requires more computational work.</p></li>
<li><p><strong>fine_tune_threshold</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#float" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a></span>) – Maximum difference from the maximum correlation to use in fine-tuning.
All labels above this threshold are used for another round of fine-tuning.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use during classification.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.6.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code> containing the <code class="docutils literal notranslate"><span class="pre">best</span></code>
label, the <code class="docutils literal notranslate"><span class="pre">scores</span></code> for each label as a nested <code class="docutils literal notranslate"><span class="pre">BiocFrame</span></code>, and the
<code class="docutils literal notranslate"><span class="pre">delta</span></code> from the best to the second-best label. Each row corresponds
to a column of <code class="docutils literal notranslate"><span class="pre">test</span></code>. The metadata contains <code class="docutils literal notranslate"><span class="pre">markers</span></code>, a list of
the markers from each pairwise comparison between labels; and <code class="docutils literal notranslate"><span class="pre">used</span></code>,
a list containing the union of markers from all comparisons.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-singler.get_classic_markers">
<span id="singler-get-classic-markers-module"></span><h2>singler.get_classic_markers module<a class="headerlink" href="#module-singler.get_classic_markers" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="singler.get_classic_markers.get_classic_markers">
<span class="sig-prename descclassname"><span class="pre">singler.get_classic_markers.</span></span><span class="sig-name descname"><span class="pre">get_classic_markers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'logcounts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_de</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/get_classic_markers.html#get_classic_markers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.get_classic_markers.get_classic_markers" title="Link to this definition">¶</a></dt>
<dd><p>Compute markers from a reference using the classic SingleR algorithm. This is typically done for reference
datasets derived from replicated bulk transcriptomic experiments.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref_data</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Any" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#list" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Any" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>]]</span>) – <p>A matrix-like object containing the log-normalized expression
values of a reference dataset.  Each column is a sample and each
row is a feature.</p>
<p>Alternatively, this can be a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.5)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>
containing a matrix-like object in one of its assays.</p>
<p>Alternatively, a list of such matrices or <code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code>
objects, typically for multiple batches of the same reference; it
is assumed that different batches exhibit at least some overlap in
their <code class="docutils literal notranslate"><span class="pre">ref_features</span></code> and <code class="docutils literal notranslate"><span class="pre">ref_labels</span></code>.</p>
</p></li>
<li><p><strong>ref_labels</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#list" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]]</span>) – <p>If <code class="docutils literal notranslate"><span class="pre">ref_data</span></code> is not a list, <code class="docutils literal notranslate"><span class="pre">ref_labels</span></code> should be a sequence
of length equal to the number of columns of <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>,
containing a label (usually a string) for each column.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">ref_data</span></code> is a list, <code class="docutils literal notranslate"><span class="pre">ref_labels</span></code> should also be a list of
the same length. Each entry should be a sequence of length equal to
the number of columns of the corresponding entry of <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>.</p>
</p></li>
<li><p><strong>ref_features</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#list" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]]</span>) – <p>If <code class="docutils literal notranslate"><span class="pre">ref_data</span></code> is not a list, <code class="docutils literal notranslate"><span class="pre">ref_features</span></code> should be a
sequence of length equal to the number of rows of <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>,
containing the feature name (usually a string) for each row.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">ref_data</span></code> is a list, <code class="docutils literal notranslate"><span class="pre">ref_features</span></code> should also be a list
of the same length. Each entry should be a sequence of length
equal to the number of rows of the corresponding entry of <code class="docutils literal notranslate"><span class="pre">ref</span></code>.</p>
</p></li>
<li><p><strong>assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Name or index of the assay of interest, if <code class="docutils literal notranslate"><span class="pre">ref</span></code> is or contains
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.5)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>
objects.</p></li>
<li><p><strong>check_missing</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to check for and remove rows with missing (NaN) values in
the reference matrices. This can be set to False if it is known
that no NaN values exist.</p></li>
<li><p><strong>num_de</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Optional" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Number of differentially expressed genes to use as markers for each
pairwise comparison between labels. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, an appropriate
number of genes is automatically determined.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use for the calculations.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Any" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Any" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#list" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>]]</span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A dictionary of dictionary of lists containing the markers for each
pairwise comparison between labels, i.e., <code class="docutils literal notranslate"><span class="pre">markers[a][b]</span></code> contains
the upregulated markers for label <code class="docutils literal notranslate"><span class="pre">a</span></code> over label <code class="docutils literal notranslate"><span class="pre">b</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="singler.get_classic_markers.number_of_classic_markers">
<span class="sig-prename descclassname"><span class="pre">singler.get_classic_markers.</span></span><span class="sig-name descname"><span class="pre">number_of_classic_markers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_labels</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/get_classic_markers.html#number_of_classic_markers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.get_classic_markers.number_of_classic_markers" title="Link to this definition">¶</a></dt>
<dd><p>Compute the number of markers to detect for a given number of labels,
using the classic SingleR marker detection algorithm.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>num_labels</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of labels.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Number of markers.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)">int</a></p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-singler.train_integrated">
<span id="singler-train-integrated-module"></span><h2>singler.train_integrated module<a class="headerlink" href="#module-singler.train_integrated" title="Link to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="singler.train_integrated.TrainedIntegratedReferences">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">singler.train_integrated.</span></span><span class="sig-name descname"><span class="pre">TrainedIntegratedReferences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_labels</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/train_integrated.html#TrainedIntegratedReferences"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.train_integrated.TrainedIntegratedReferences" title="Link to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Object containing integrated references, typically constructed by
<a class="reference internal" href="#singler.train_integrated.train_integrated" title="singler.train_integrated.train_integrated"><code class="xref py py-meth docutils literal notranslate"><span class="pre">train_integrated()</span></code></a>.</p>
<dl class="py property">
<dt class="sig sig-object py" id="singler.train_integrated.TrainedIntegratedReferences.reference_labels">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reference_labels</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#list" title="(in Python v3.12)"><span class="pre">list</span></a></em><a class="headerlink" href="#singler.train_integrated.TrainedIntegratedReferences.reference_labels" title="Link to this definition">¶</a></dt>
<dd><p>List of lists containing the names of the labels for each reference.</p>
<p>Each entry corresponds to a reference in <code class="xref py py-attr docutils literal notranslate"><span class="pre">reference_names</span></code>,
if <code class="docutils literal notranslate"><span class="pre">reference_names</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="singler.train_integrated.train_integrated">
<span class="sig-prename descclassname"><span class="pre">singler.train_integrated.</span></span><span class="sig-name descname"><span class="pre">train_integrated</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">test_features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_prebuilt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">warn_lost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/train_integrated.html#train_integrated"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.train_integrated.train_integrated" title="Link to this definition">¶</a></dt>
<dd><p>Build a set of integrated references for classification of a test dataset.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>test_features</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Sequence of features for the test dataset.</p></li>
<li><p><strong>ref_prebuilt</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#list" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>[<a class="reference internal" href="#singler.train_single.TrainedSingleReference" title="singler.train_single.TrainedSingleReference"><code class="xref py py-class docutils literal notranslate"><span class="pre">TrainedSingleReference</span></code></a>]</span>) – List of prebuilt references, typically created by calling
<a class="reference internal" href="#singler.train_single.train_single" title="singler.train_single.train_single"><code class="xref py py-meth docutils literal notranslate"><span class="pre">train_single()</span></code></a>.</p></li>
<li><p><strong>warn_lost</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to emit a warning if the markers for each reference are not
all present in all references.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#singler.train_integrated.TrainedIntegratedReferences" title="singler.train_integrated.TrainedIntegratedReferences"><code class="xref py py-class docutils literal notranslate"><span class="pre">TrainedIntegratedReferences</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Integrated references for classification with
<code class="xref py py-meth docutils literal notranslate"><span class="pre">classify_integrated()</span></code>.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-singler.train_single">
<span id="singler-train-single-module"></span><h2>singler.train_single module<a class="headerlink" href="#module-singler.train_single" title="Link to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="singler.train_single.TrainedSingleReference">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">singler.train_single.</span></span><span class="sig-name descname"><span class="pre">TrainedSingleReference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">full_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">full_label_codes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">markers</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/train_single.html#TrainedSingleReference"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.train_single.TrainedSingleReference" title="Link to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>A prebuilt reference object, typically created by
<a class="reference internal" href="#singler.train_single.train_single" title="singler.train_single.train_single"><code class="xref py py-meth docutils literal notranslate"><span class="pre">train_single()</span></code></a>. This is intended for
advanced users only and should not be serialized.</p>
<dl class="py property">
<dt class="sig sig-object py" id="singler.train_single.TrainedSingleReference.features">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">features</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#list" title="(in Python v3.12)"><span class="pre">list</span></a></em><a class="headerlink" href="#singler.train_single.TrainedSingleReference.features" title="Link to this definition">¶</a></dt>
<dd><p>The universe of features known to this reference.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="singler.train_single.TrainedSingleReference.labels">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">labels</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><span class="pre">Sequence</span></a></em><a class="headerlink" href="#singler.train_single.TrainedSingleReference.labels" title="Link to this definition">¶</a></dt>
<dd><p>Unique labels in this reference.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="singler.train_single.TrainedSingleReference.marker_subset">
<span class="sig-name descname"><span class="pre">marker_subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">indices_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/train_single.html#TrainedSingleReference.marker_subset"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.train_single.TrainedSingleReference.marker_subset" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>indices_only</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to return the markers as indices
into <a class="reference internal" href="#singler.train_single.TrainedSingleReference.features" title="singler.train_single.TrainedSingleReference.features"><code class="xref py py-attr docutils literal notranslate"><span class="pre">features</span></code></a>, or as a list of feature identifiers.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#list" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>]</span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>If <code class="docutils literal notranslate"><span class="pre">indices_only</span> <span class="pre">=</span> <span class="pre">False</span></code>, a list of feature identifiers for the markers.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">indices_only</span> <span class="pre">=</span> <span class="pre">True</span></code>, a NumPy array containing the integer indices of
features in <code class="docutils literal notranslate"><span class="pre">features</span></code> that were chosen as markers.</p>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="singler.train_single.TrainedSingleReference.markers">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">markers</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#dict" title="(in Python v3.12)"><span class="pre">dict</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Any" title="(in Python v3.12)"><span class="pre">Any</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#dict" title="(in Python v3.12)"><span class="pre">dict</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Any" title="(in Python v3.12)"><span class="pre">Any</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#list" title="(in Python v3.12)"><span class="pre">list</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#singler.train_single.TrainedSingleReference.markers" title="Link to this definition">¶</a></dt>
<dd><p>Markers for every pairwise comparison between labels.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="singler.train_single.TrainedSingleReference.num_labels">
<span class="sig-name descname"><span class="pre">num_labels</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/train_single.html#TrainedSingleReference.num_labels"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.train_single.TrainedSingleReference.num_labels" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Number of unique labels in this reference.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="singler.train_single.TrainedSingleReference.num_markers">
<span class="sig-name descname"><span class="pre">num_markers</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/train_single.html#TrainedSingleReference.num_markers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.train_single.TrainedSingleReference.num_markers" title="Link to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Number of markers to be used for classification. This is the
same as the size of the array from <a class="reference internal" href="#singler.train_single.TrainedSingleReference.marker_subset" title="singler.train_single.TrainedSingleReference.marker_subset"><code class="xref py py-meth docutils literal notranslate"><span class="pre">marker_subset()</span></code></a>.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="singler.train_single.train_single">
<span class="sig-prename descclassname"><span class="pre">singler.train_single.</span></span><span class="sig-name descname"><span class="pre">train_single</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_features=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type='logcounts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">restrict_to=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_missing=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">markers=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_method='classic'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_de=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_args={}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aggregate=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aggregate_args={}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_parameters=&lt;knncolle.vptree.VptreeParameters</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads=1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/train_single.html#train_single"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.train_single.train_single" title="Link to this definition">¶</a></dt>
<dd><p>Build a single reference dataset in preparation for classification.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref_data</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Any" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span>) – <p>A matrix-like object where rows are features, columns are
reference profiles, and each entry is the expression value.
If <code class="docutils literal notranslate"><span class="pre">markers</span></code> is not provided, expression should be normalized
and log-transformed in preparation for marker prioritization via
differential expression analyses. Otherwise, any expression values
are acceptable as only the ranking within each column is used.</p>
<p>Alternatively, a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.5)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>
containing such a matrix in one of its assays.</p>
</p></li>
<li><p><strong>ref_labels</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Sequence of labels for each reference profile, i.e., column in <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>.</p></li>
<li><p><strong>ref_features</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a></span>) – Sequence of identifiers for each feature, i.e., row in <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>.</p></li>
<li><p><strong>test_features</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Optional" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]</span>) – Sequence of identifiers for each feature in the test dataset.</p></li>
<li><p><strong>assay_type</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Assay containing the expression matrix,
if <code class="docutils literal notranslate"><span class="pre">ref_data</span></code> is a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.5.5)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>.</p></li>
<li><p><strong>check_missing</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether to check for and remove rows with missing (NaN) values
from <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>.</p></li>
<li><p><strong>restrict_to</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Union" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#set" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">set</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/constants.html#None" title="(in Python v3.12)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span>) – Subset of available features to restrict to. Only features in
<code class="docutils literal notranslate"><span class="pre">restrict_to</span></code> will be used in the reference building. If
<code class="docutils literal notranslate"><span class="pre">None</span></code>, no restriction is performed.</p></li>
<li><p><strong>markers</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Optional" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Any" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Any" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>, <a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]]]</span>) – Upregulated markers for each pairwise comparison between labels.
Specifically, <code class="docutils literal notranslate"><span class="pre">markers[a][b]</span></code> should be a sequence of features
that are upregulated in <code class="docutils literal notranslate"><span class="pre">a</span></code> compared to <code class="docutils literal notranslate"><span class="pre">b</span></code>. All such features
should be present in <code class="docutils literal notranslate"><span class="pre">features</span></code>, and all labels in <code class="docutils literal notranslate"><span class="pre">labels</span></code>
should have keys in the inner and outer dictionaries.</p></li>
<li><p><strong>marker_method</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Literal" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'classic'</span></code>, <code class="docutils literal notranslate"><span class="pre">'auc'</span></code>, <code class="docutils literal notranslate"><span class="pre">'cohens_d'</span></code>]</span>) – Method to identify markers from each pairwise comparisons between labels in <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>.
If <code class="docutils literal notranslate"><span class="pre">classic</span></code>, we call <a class="reference internal" href="#singler.get_classic_markers.get_classic_markers" title="singler.get_classic_markers.get_classic_markers"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_classic_markers()</span></code></a>.
If <code class="docutils literal notranslate"><span class="pre">auc</span></code> or <code class="docutils literal notranslate"><span class="pre">cohens_d</span></code>, we call <code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers()</span></code>.
Only used if <code class="docutils literal notranslate"><span class="pre">markers</span></code> is not supplied.</p></li>
<li><p><strong>num_de</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Optional" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span>) – Number of differentially expressed genes to use as markers for each pairwise comparison between labels.
If <code class="docutils literal notranslate"><span class="pre">None</span></code> and <code class="docutils literal notranslate"><span class="pre">marker_method</span> <span class="pre">=</span> <span class="pre">&quot;classic&quot;</span></code>, an appropriate number of genes is determined by <a class="reference internal" href="#singler.get_classic_markers.get_classic_markers" title="singler.get_classic_markers.get_classic_markers"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_classic_markers()</span></code></a>.
Otherwise, it is set to 10.
Only used if <code class="docutils literal notranslate"><span class="pre">markers</span></code> is not supplied.</p></li>
<li><p><strong>marker_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Further arguments to pass to the chosen marker detection method.
If <code class="docutils literal notranslate"><span class="pre">marker_method</span> <span class="pre">=</span> <span class="pre">&quot;classic&quot;</span></code>, this is <a class="reference internal" href="#singler.get_classic_markers.get_classic_markers" title="singler.get_classic_markers.get_classic_markers"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_classic_markers()</span></code></a>, otherwise it is <code class="xref py py-func docutils literal notranslate"><span class="pre">score_markers()</span></code>.
Only used if <code class="docutils literal notranslate"><span class="pre">markers</span></code> is not supplied.</p></li>
<li><p><strong>aggregate</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span>) – Whether the reference dataset should be aggregated to pseudo-bulk samples for speed, see <a class="reference internal" href="#singler.aggregate_reference.aggregate_reference" title="singler.aggregate_reference.aggregate_reference"><code class="xref py py-func docutils literal notranslate"><span class="pre">aggregate_reference()</span></code></a> for details.</p></li>
<li><p><strong>aggregate_args</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/stdtypes.html#dict" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span>) – Further arguments to pass to <a class="reference internal" href="#singler.aggregate_reference.aggregate_reference" title="singler.aggregate_reference.aggregate_reference"><code class="xref py py-func docutils literal notranslate"><span class="pre">aggregate_reference()</span></code></a> when <code class="docutils literal notranslate"><span class="pre">aggregate</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p></li>
<li><p><strong>nn_parameters</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/typing.html#typing.Optional" title="(in Python v3.12)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Parameters</span></code>]</span>) – Algorithm for constructing the neighbor search index, used to
compute scores during classification.</p></li>
<li><p><strong>num_threads</strong> (<span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3.12/library/functions.html#int" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span>) – Number of threads to use for reference building.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#singler.train_single.TrainedSingleReference" title="singler.train_single.TrainedSingleReference"><code class="xref py py-class docutils literal notranslate"><span class="pre">TrainedSingleReference</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The pre-built reference, ready for use in downstream methods like
<code class="xref py py-meth docutils literal notranslate"><span class="pre">classify_single()</span></code>.</p>
</dd>
</dl>
</dd></dl>

</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          <a class="prev-page" href="modules.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">singler</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2023, Aaron Lun
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">singler package</a><ul>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-singler.aggregate_reference">singler.aggregate_reference module</a><ul>
<li><a class="reference internal" href="#singler.aggregate_reference.aggregate_reference"><code class="docutils literal notranslate"><span class="pre">aggregate_reference()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-singler.annotate_integrated">singler.annotate_integrated module</a><ul>
<li><a class="reference internal" href="#singler.annotate_integrated.annotate_integrated"><code class="docutils literal notranslate"><span class="pre">annotate_integrated()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-singler.annotate_single">singler.annotate_single module</a><ul>
<li><a class="reference internal" href="#singler.annotate_single.annotate_single"><code class="docutils literal notranslate"><span class="pre">annotate_single()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-singler.classify_integrated">singler.classify_integrated module</a><ul>
<li><a class="reference internal" href="#singler.classify_integrated.classify_integrated"><code class="docutils literal notranslate"><span class="pre">classify_integrated()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-singler.classify_single">singler.classify_single module</a><ul>
<li><a class="reference internal" href="#singler.classify_single.classify_single"><code class="docutils literal notranslate"><span class="pre">classify_single()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-singler.get_classic_markers">singler.get_classic_markers module</a><ul>
<li><a class="reference internal" href="#singler.get_classic_markers.get_classic_markers"><code class="docutils literal notranslate"><span class="pre">get_classic_markers()</span></code></a></li>
<li><a class="reference internal" href="#singler.get_classic_markers.number_of_classic_markers"><code class="docutils literal notranslate"><span class="pre">number_of_classic_markers()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-singler.train_integrated">singler.train_integrated module</a><ul>
<li><a class="reference internal" href="#singler.train_integrated.TrainedIntegratedReferences"><code class="docutils literal notranslate"><span class="pre">TrainedIntegratedReferences</span></code></a><ul>
<li><a class="reference internal" href="#singler.train_integrated.TrainedIntegratedReferences.reference_labels"><code class="docutils literal notranslate"><span class="pre">TrainedIntegratedReferences.reference_labels</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#singler.train_integrated.train_integrated"><code class="docutils literal notranslate"><span class="pre">train_integrated()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-singler.train_single">singler.train_single module</a><ul>
<li><a class="reference internal" href="#singler.train_single.TrainedSingleReference"><code class="docutils literal notranslate"><span class="pre">TrainedSingleReference</span></code></a><ul>
<li><a class="reference internal" href="#singler.train_single.TrainedSingleReference.features"><code class="docutils literal notranslate"><span class="pre">TrainedSingleReference.features</span></code></a></li>
<li><a class="reference internal" href="#singler.train_single.TrainedSingleReference.labels"><code class="docutils literal notranslate"><span class="pre">TrainedSingleReference.labels</span></code></a></li>
<li><a class="reference internal" href="#singler.train_single.TrainedSingleReference.marker_subset"><code class="docutils literal notranslate"><span class="pre">TrainedSingleReference.marker_subset()</span></code></a></li>
<li><a class="reference internal" href="#singler.train_single.TrainedSingleReference.markers"><code class="docutils literal notranslate"><span class="pre">TrainedSingleReference.markers</span></code></a></li>
<li><a class="reference internal" href="#singler.train_single.TrainedSingleReference.num_labels"><code class="docutils literal notranslate"><span class="pre">TrainedSingleReference.num_labels()</span></code></a></li>
<li><a class="reference internal" href="#singler.train_single.TrainedSingleReference.num_markers"><code class="docutils literal notranslate"><span class="pre">TrainedSingleReference.num_markers()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#singler.train_single.train_single"><code class="docutils literal notranslate"><span class="pre">train_single()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../_static/documentation_options.js?v=3f1b9271"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/furo.js?v=5fa4622c"></script>
    </body>
</html>